File: java/vertx/lib/src/main/java/reis/polyglot/lib/apiResponses/ApiResponse.java
package reis.polyglot.lib.apiResponses;

import reis.polyglot.lib.exceptions.ApplicationException;

public abstract class ApiResponse<T extends JsonSerializable> {
    boolean success;

    int statusCode;

    T data;

    ApplicationException error;

    public ApiResponse(int statusCode, T data) {
        this.data = data;
        this.error = null;
        this.statusCode = statusCode;
        this.success = true;
    }

    public ApiResponse(int statusCode, ApplicationException error) {
        this.data = null;
        this.error = error;
        this.statusCode = statusCode;
        this.success = false;
    }
}

File: servers/javascript/express/src/api/v1/lib/responses/api-response.ts
import { Response } from 'express';
import { ApplicationError } from '@lib/exceptions/application-exception';

export abstract class ApiResponse {
    constructor(
        public readonly statusCode: number,
        public readonly success: boolean,
    ) {}

    protected formatAndSend(
        res: Response,
        error: ApplicationError | null,
        data: any,
    ) {
        const body = {
            data,
            error,
            success: this.success,
        };
        res.status(this.statusCode).json(body);
    }

    abstract send(res: Response): void;
}
